server/main.py

```diff
@@ -218,38 +214,42 @@ async def room(websocket: WebSocket) -> None:
     """
     client = Client(websocket)
     await client.accept()
+
     try:
-        initial_data = ConnectionData(**await client.receive())
+        initial_event = await client.receive()
     except WebSocketDisconnect:
         return
 
-    room_code = initial_data.data.room_code
-    if initial_data.type == "connect":
-        match initial_data.data.connection_type:
-            case "create":
-                manager.create_room(client, room_code)
-            case "join":
-                try:
-                    manager.join_room(client, room_code)
-                except RoomNotFoundError as e:
-                    # Send off to frontend to handle
-                    await client.send(
-                        {
-                            "type": "error",
-                            "data": {
-                                "message": e.message,
-                            },
-                            "status_code": StatusCode.ROOM_NOT_FOUND,
-                        }
-                    )
-                    await client.close()
-                    return
-            case _:
-                raise NotImplementedError
+    if initial_event.type != EventType.CONNECT:
+        return
+
+    initial_data: ConnectData = initial_event.data
+    room_code = initial_data.room_code
+
+    try:
+        ConnectionManager.connect(client, room_code, initial_data.connection_type)
+    except RoomNotFoundError as e:
+        await client.send(e.data)
+        await client.close()
+        return
+
+    await manager.broadcast(initial_event, room_code)
 
     try:
         while True:
-            data = await client.receive()
-            await manager.broadcast(data, room_code, sender=client)
+            event = await client.receive()
+
+            if event.type == EventType.REPLACE:
+                manager.update_code_cache(room_code, event.data)
+
+            await manager.broadcast(event, room_code)
     except WebSocketDisconnect:
+        await manager.broadcast(
+            EventResponse(
+                type=EventType.DISCONNECT,
+                data=DisconnectData(username=initial_data.username),
+                status_code=StatusCode.SUCCESS,
+            ),
+            room_code,
+        )
```

Comment:
```py
room_code,
sender=cleint
)
```
This fixes the RuntimeError so that it doesn't broadcast a message to itself if they have already disconnected

Reply:
Fixed in 3b9af74630826320c39bdfe2a6b73d31a2f7b25d.